### PR TITLE
add bf16 and change header file include path

### PR DESCRIPTION
**Pull request by guangyey**
Fixes #ISSUE_NUMBER

### CODE REVIEW

File: c10/util/BFloat16-inl.h [hunk 1]
```diff
@@ -8,6 +8,15 @@ C10_CLANG_DIAGNOSTIC_PUSH()
 C10_CLANG_DIAGNOSTIC_IGNORE("-Wimplicit-int-float-conversion")
 #endif
 
+#ifdef SYCL_LANGUAGE_VERSION
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp> // for SYCL 2020
+#elif __has_include(<CL/sycl.hpp>)
```
Comment (gujinghui) [inline c10/util/BFloat16-inl.h:14]: we cannot use __has_include as long as PyTorch does not build with C++17 as default.

File: c10/util/BFloat16-inl.h [hunk 2]
```diff
@@ -8,6 +8,15 @@ C10_CLANG_DIAGNOSTIC_PUSH()
...
+#include <CL/sycl.hpp> // for SYCL 1.2.1
+#endif
+#include <ext/oneapi/bfloat16.hpp>
```
Comment (chengjunlu) [inline c10/util/BFloat16-inl.h:17]: Is this Intel specific? 
We need to wrap the Intel specific header with Intel macro.

Comment (gujinghui) [inline c10/util/BFloat16-inl.h:17]: This is a extention of SYCL from DPC++ compiler.
We should find some DPC++ compiler macro to control it.

File: c10/util/BFloat16-inl.h [hunk 3]
```diff
@@ -16,6 +25,8 @@ inline C10_HOST_DEVICE BFloat16::BFloat16(float value)
 #if defined(CUDA_VERSION) && CUDA_VERSION >= 11000 && \
     defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 800
       x(__bfloat16_as_ushort(__float2bfloat16(value)))
+#elif defined(__SYCL_DEVICE_ONLY__)
+      x = sycl::bit_cast<uint16_t>(sycl::ext::oneapi::bfloat16(value));
```
Comment (chengjunlu) [inline c10/util/BFloat16-inl.h:29]: Same problem.
I don't think the "oneapi" namespace is public in SYCL. It is Intel extension.

File: c10/util/BFloat16-inl.h (hunk 2 above)
Comment (guangyey) [reply c10/util/BFloat16-inl.h:18]: I find that #if defined(SYCL_LANGUAGE_VERSION) && defined (__INTEL_LLVM_COMPILER) is used for code specific for DPC++ compiler. And I find another macro **SYCL_EXT_ONEAPI_BFLOAT16** in feature_test.hpp.
which one is better?

Comment (guangyey) [reply c10/util/BFloat16-inl.h:18]: I think SYCL_EXT_ONEAPI_BFLOAT16 is not a good choice. It is not a stable macro.

Comment (chengjunlu) [reply c10/util/BFloat16-inl.h:18]: Only find a little Intel MACRO in clang front end.

https://github.com/intel-innersource/applications.compilers.source.llvm-project-shared/blob/71a7fac5f1ccf7f337bc2d397f752c44fdefee36/clang/lib/Frontend/InitPreprocessor.cpp#L768

None is good enough to distinguish the extension capability of oneapi.

Let's try to use the "__INTEL_CLANG_COMPILER" and its version number at first. 

And we also need to double confirm that it is properly used for our purpose with compiler team. Please create a JIRA to DPCPP CMPLR team.

Comment (chengjunlu) [reply c10/util/BFloat16-inl.h:18]: The best we can ask the DPCPP team to add the bf16 header into the sycl.hpp.

https://github.com/intel-innersource/applications.compilers.source.llvm-project-shared/blob/71a7fac5f1ccf7f337bc2d397f752c44fdefee36/sycl/include/sycl/sycl.hpp#L63

Please file a JIRA to discuss about it with DPCPP team.

File: c10/util/BFloat16-inl.h [hunk 4]
```diff
@@ -8,6 +8,15 @@ C10_CLANG_DIAGNOSTIC_PUSH()
 C10_CLANG_DIAGNOSTIC_IGNORE("-Wimplicit-int-float-conversion")
 #endif
 
+#if defined(CL_SYCL_LANGUAGE_VERSION)
```
Comment (guangyey) [inline c10/util/BFloat16-inl.h:11]: refer to https://github.com/intel/llvm/blob/59dd38795c7557a779ca38cd27df21a40c7b3c45/clang/lib/Basic/Version.cpp#L129
In my option, SYCL_LANGUAGE_VERSION is defined in both SYCL 1.2.1 and SYCL 2020. But only CL_SYCL_LANGUAGE_VERSION is defined in SYCL 1.2.1.
So we should check CL_SYCL_LANGUAGE_VERSION first for SYCL 1.2.1. If it is not defined then check SYCL_LANGUAGE_VERSION for SYCL 2020.
Is it right?

File: c10/util/BFloat16-inl.h [hunk 5]
```diff
@@ -8,6 +8,15 @@ C10_CLANG_DIAGNOSTIC_PUSH()
...
+#include <CL/sycl.hpp> // for SYCL 1.2.1
+#elif defined(SYCL_LANGUAGE_VERSION)
+#include <sycl/sycl.hpp> // for SYCL 2020
+#endif
+#if defined(SYCL_LANGUAGE_VERSION) && defined(__INTEL_LLVM_COMPILER)
```
Comment (guangyey) [inline c10/util/BFloat16-inl.h:16]: refer to https://www.intel.com/content/www/us/en/developer/articles/technical/use-predefined-macros-for-specific-code-for-intel-dpcpp-compiler-intel-cpp-compiler-intel-cpp-compiler-classic.html
We can use #if defined(SYCL_LANGUAGE_VERSION) && defined(__INTEL_LLVM_COMPILER) for code specific for DPC++ compiler. It guarantees that /bfloat16.hpp is included in the Intel SYCL extension implemented from the Intel DPC++ compiler.

**Review by gujinghui — APPROVED** [review]

File: c10/util/BFloat16-inl.h [hunk 6]
```diff
@@ -8,6 +8,15 @@ C10_CLANG_DIAGNOSTIC_PUSH()
 C10_CLANG_DIAGNOSTIC_IGNORE("-Wimplicit-int-float-conversion")
 #endif
 
+#if defined(CL_SYCL_LANGUAGE_VERSION)
+#include <CL/sycl.hpp> // for SYCL 1.2.1
+#elif defined(SYCL_LANGUAGE_VERSION)
+#include <sycl/sycl.hpp> // for SYCL 2020
```
Comment (gujinghui) [inline c10/util/BFloat16-inl.h:14]: Should be as below, as if no bfloat16 header the sycl header is useless here.

#if defined(SYCL_LANGUAGE_VERSION) && defined(__INTEL_LLVM_COMPILER)
#if defined(CL_SYCL_LANGUAGE_VERSION)
#include <CL/sycl.hpp> // for SYCL 1.2.1
#elif defined(SYCL_LANGUAGE_VERSION)
#include <sycl/sycl.hpp> // for SYCL 2020
#endif
#include <ext/oneapi/bfloat16.hpp>
#endif

Comment (guangyey) [reply c10/util/BFloat16-inl.h:14]: OK

**Pull request closed without merging**